src/services/route.rs

```diff
@@ -0,0 +1,257 @@
+//! This module contains the implementation of a service for
+//! setting the url and responding to changes to the url
+//! that are initiated by the browser..
+
+use stdweb::web::History;
+use stdweb::web::Location;
+use stdweb::web::window;
+//use stdweb::web::Window;
+//use stdweb;
+use stdweb::Value;
+use stdweb::web::EventListenerHandle;
+use stdweb::web::event::PopStateEvent;
+use stdweb::web::IEventTarget;
+use stdweb::unstable::TryFrom;
+use html::Callback;
+
+use html::Env;
+use html::Component;
+use url_lib::{Url};
+use std::ops::Add;
+
+
+type RouteDidChange = bool;
+
+/// An alias for `Result<RouteInfo, RoutingError>`.
+pub type RouteResult = Result<RouteInfo, RoutingError>;
+
+/// Service used for routing
+pub struct RouteService {
+    history: History,
+    location: Location,
+    event_listener: Option<EventListenerHandle>,
+    callback: Option<Callback<RouteResult>>
+}
+
+/// A subset of the url crate's Url object that can be passed
+/// to crate consumers to deal with routing.
+#[derive(Clone, PartialEq, Debug)]
+pub struct RouteInfo {
```

Comment:
I strongly prefer that this be only two items:

- `state: Option<Value>`
- `location: Location`

The `Location` type is then just the _resolved_ items in:
- `js!{return window.location}`
- OR `js!{return new URL(@{href.to_string()})}`

i.e. _we_ will handle any thrown `SecurityError`s and panic/log there.

Reply:
The main benefit is that we do not require an extra library, but also we are standards compliant with browsers.

Reply:
I disagree that `RouteInfo` should require holding onto a `Location` struct. I think that `RouteInfo` should only contain information essential to routing, which excludes information in the origin. This implies that it should contain 
*  `path: String` or `path_segments: Vec<String>
*  `query: Option<String>`
*  `fragment: Option<String>`
*  `state: Option<T: JsSerialize>`

I do agree that the Location API, once StdWeb adds support for it, should be used to get info about the current route during the handling of a `PopStateEvent`. I think my current approach is misguided in storing the route string in the history's state, and then parsing that into a `RouteInfo`. Instead, when the event is fired, I should get the RouteInfo from the location API (currently by parsing the href via `url`, in the future, by using stdweb's `Location`'s fields), and try to attach the state if it was `Some`.

I still do want to use the `url` crate because my routing implementation uses it to validate routes passed to it by crate users, and I trust it to perform better than anything I could ever write.

Reply:
So, my `push_state` ended up looking like this:

```
    /// Set the state of the history, including the url.
    ///
    /// This will only trigger a state change in the frontend if `emit == false`
    pub fn push_state(&self, emit: bool, state: Value, title: &str, url: Option<&str>) {
        let route_fn = match self.route_fn {
            Some(r) => r,
            None => panic!("Attempted to set_state without initializing router"),
        };
        let url = match url {
            Some(url) => url.to_string(),
            None => self.window.location().unwrap().href().unwrap(),
        };
        self.history.push_state(state.clone(), title, Some(&url));
        if update {
            let info = RouteInfo {
                location: parse_location(&url),
                state: state,
            };
            self.callback.as_ref().unwrap().emit(info)
        }
    }
```

That `parse_location` basically just calls `js!{ return new Url(url); }`.

So... I ended up parsing the url anyway! I'm starting to think that doing it through a library is probably worth it.

This brings up an interesting point... _none_ of what we are doing here _needs_ to be in the `yew` crate itself. We could make our own `yew_route` crate! This would allow us to experiment and include whatever dependencies we might want.

### Conclusion
I'm coming around to doing the url parsing in rust. I also agree that having type safety on the URL types (via a macro in the future) is worth pursuing.

However, I disagree with your layout. Why wouldn't `RouteInfo` just be two fields, `state` and `url`, where [`url` is the `Url` type](https://docs.rs/url/1.7.0/url/struct.Url.html)?

Reply:
One thing I want to make sure we support is the ability to `set_state` to a partial URL, i.e. just the hash/query.

Another possibility is to require the user to hold onto their initial `Url` "base" and use things like [`set_query`](https://docs.rs/url/1.7.0/url/struct.Url.html#method.set_query) to set the query they want, etc.

Setting the query based on a base-url is currently annoying, so I opened servo/rust-url#447